src/core/DY.Core.Bytes.fst

```diff
@@ -1062,15 +1070,37 @@ let rec bytes_invariant_later #cinvs tr1 tr2 msg =
     bytes_invariant_later tr1 tr2 nonce;
     bytes_invariant_later tr1 tr2 msg;
     bytes_invariant_later tr1 tr2 ad;
-    FStar.Classical.forall_intro_4 (FStar.Classical.move_requires_4 (aead_pred.pred_later tr1 tr2))
+    introduce (exists key_usg. key `has_usage tr1` key_usg /\ AeadKey? key_usg /\ aead_pred.pred tr1 key_usg key nonce msg ad) ==> 
+      (exists key_usg. key `has_usage tr2` key_usg /\ aead_pred.pred tr2 key_usg key nonce msg ad)
+    with _. (
+      assert(exists key_usg. key `has_usage tr1` key_usg /\ AeadKey? key_usg /\ aead_pred.pred tr1 key_usg key nonce msg ad);
+      eliminate exists key_usg. key `has_usage tr1` key_usg /\ AeadKey? key_usg /\ aead_pred.pred tr1 key_usg key nonce msg ad
+      returns (exists key_usg. key `has_usage tr2` key_usg /\ aead_pred.pred tr2 key_usg key nonce msg ad)
+      with _. (
+        aead_pred.pred_later tr1 tr2 key_usg key nonce msg ad;
+        assert(exists key_usg. key `has_usage tr2` key_usg /\ aead_pred.pred tr2 key_usg key nonce msg ad);
+        ()
+      )
+    )
   )
   | Pk sk ->
     bytes_invariant_later tr1 tr2 sk
   | PkeEnc pk nonce msg -> (
     bytes_invariant_later tr1 tr2 pk;
     bytes_invariant_later tr1 tr2 nonce;
     bytes_invariant_later tr1 tr2 msg;
-    FStar.Classical.forall_intro_2 (FStar.Classical.move_requires_2 (pke_pred.pred_later tr1 tr2))
+    introduce (exists sk_usg. pk `has_sk_usage tr1` sk_usg /\ PkeKey? sk_usg /\ pke_pred.pred tr1 sk_usg pk msg) ==>
+      (exists sk_usg. pk `has_sk_usage tr2` sk_usg /\ pke_pred.pred tr2 sk_usg pk msg)
+    with _. (
+      assert(exists sk_usg. pk `has_sk_usage tr1` sk_usg /\ PkeKey? sk_usg /\ pke_pred.pred tr1 sk_usg pk msg);
+      eliminate exists sk_usg. pk `has_sk_usage tr1` sk_usg /\ PkeKey? sk_usg /\ pke_pred.pred tr1 sk_usg pk msg
+      returns (exists sk_usg. pk `has_sk_usage tr2` sk_usg /\ pke_pred.pred tr2 sk_usg pk msg)
+      with _. (
+        pke_pred.pred_later tr1 tr2 sk_usg pk msg;
+        assert(exists sk_usg. pk `has_sk_usage tr2` sk_usg /\ pke_pred.pred tr2 sk_usg pk msg);
+        ()
+      )
+    )
```

Comment:
Here, it seems like there are few enough arguments that `FStar.Classical.forall_intro_3 (FStar.Classical.move_requires_3 (pke_pred.pred_later tr1 tr2))` would work (likewise for signatures). The lack of `_5` versions of these makes that not work for AEAD, however.

It also seems like there might be some way to unify these proofs, which all have the same structure, but it's not immediately obvious to me.

Reply:
I tried to make it work with `FStar.Classical.forall_intro_3`, but it does not work for me. Does it work for you?

Reply:
I haven't checked out this branch locally to test yet, no. I can take a look at this (and probably some of the other places where proofs got much bulkier) later today, though.